### PR TITLE
Break error intersection analysis in to two phases

### LINKED ISSUE #27758: Unknown type errors with type definitions of intersection types
**Description:**
$title. 

At least the change in type reference seems to be due to the recently introduced `org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter#intersectionTypes` not working with `org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter#unresolvedTypes`. 

**Steps to reproduce:**
```ballerina
type X error<record { int i?; }> & error<record { int i; }>;
type Y readonly & record { int i; };

type Z record {
    *Y; // previously "incompatible types: 'Y' is not a record"

    X x; // newly introduced
    Y y; // worked previously
};
```

```cmd
$ ballerina run qwe.bal 

Compiling source
        qwe.bal
ERROR [qwe.bal:(5:6,5:7)] unknown type 'Y'
ERROR [qwe.bal:(7:5,7:6)] unknown type 'X'
ERROR [qwe.bal:(8:5,8:6)] unknown type 'Y'
error: compilation contains errors
```

**Affected Versions:**
Current master

**Related PR:**
https://github.com/ballerina-platform/ballerina-lang/pull/26770

### PR DESCRIPTION
**Pull request by dulvinw**
Fixes #27758

After the record fields are symbol entered, start the second phase of error intersection.
Once the intersection is calculated, populate the information of the placeholder error type

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/intersection/IntersectionTypeTest.java [hunk 1]
```diff
@@ -17,6 +17,7 @@
  */
 package org.ballerinalang.test.types.intersection;
 
+import org.apache.commons.lang3.builder.ToStringExclude;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/intersection/IntersectionTypeTest.java:20]: Unused/unnecessary import?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 2]
```diff
@@ -1323,11 +1331,35 @@ private void populateTypeIds(BErrorType effectiveType, BLangIntersectionTypeNode
...
+        intersectionType.effectiveType = alreadyDefinedErrorType;
+
+        addErrorTypeDefinition(alreadyDefinedErrorType, env, typeDefinition.pos);
+    }
+
+    private void addErrorTypeDefinition(BErrorType errorType, SymbolEnv pkgEnv, Location pos) {
+
+        BTypeSymbol errorTSymbol = errorType.tsymbol;
+
+        // Create error type definition
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:1357]: Do we need this comment?

**Review by MaryamZi — COMMENTED** [review]
Shall we add a test for the record part in https://github.com/ballerina-platform/ballerina-lang/issues/27758 too?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 3]
```diff
@@ -1323,11 +1331,32 @@ private void populateTypeIds(BErrorType effectiveType, BLangIntersectionTypeNode
...
         BErrorType errorType = (BErrorType) intersectionType.effectiveType;
 
-        defineErrorType(errorType, pos);
+        alreadyDefinedErrorType.typeIdSet = errorType.typeIdSet;
+        alreadyDefinedErrorType.detailType = errorType.detailType;
+        alreadyDefinedErrorType.flags = errorType.flags;
+        alreadyDefinedErrorType.name = errorType.name;
+        intersectionType.effectiveType = alreadyDefinedErrorType;
+
+        addErrorTypeDefinition(alreadyDefinedErrorType, env, typeDefinition.pos);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:1350]: There is already a type definition for this error right? Shouldn't we update that or replace it altogether?

File: tests/jballerina-unit-test/src/test/resources/test-src/types/intersection/error_intersection_type_negative.bal [hunk 4]
```diff
@@ -56,3 +56,12 @@ function testRecordAndMapIntersection() {
     var err = error IntersectionErrorFour("message", x = "x", z = 10);
     DistinctErrorIntersection err2 = error IntersectionErrorFour("message", x = "x");
 }
+
+type X error<record { int i?; }> & error<record { int i; }>;
+type Y readonly & record { int i; };
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/intersection/error_intersection_type_negative.bal:61]: Shall we move the record scenarios to non-error intersection test? Shall we also add a positive test separately?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java (hunk 3 above)
Comment (dulvinw) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:1350]: No. We have only created the ErrorType and Symbol before. Hence adding the definition now

File: tests/jballerina-unit-test/src/test/resources/test-src/types/intersection/error_intersection_type_negative.bal (hunk 4 above)
Comment (dulvinw) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/intersection/error_intersection_type_negative.bal:61]: Positive test as in? For record intersections?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 5]
```diff
@@ -1006,6 +1006,14 @@ private void defineTypeNodes(List<BLangNode> typeDefs, SymbolEnv env) {
         defineTypeNodes(unresolvedTypes, env);
     }
 
+    private void populateUndefinedErrorIntersection(BLangTypeDefinition typeDef, SymbolEnv env) {
+        BErrorType intersectionErrorType = types.createErrorType(null, Flags.PUBLIC, env);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:1010]: Shouldn't the public flag come from the original definition?

File: tests/jballerina-unit-test/src/test/resources/test-src/types/intersection/test_intersection_type.bal [hunk 6]
```diff
@@ -17,7 +17,7 @@
 function testIntersectionTypes() {
     testIntersectionWithMemberTypeDefinedAfter();
     testIntersectionInUnion();
-    testComplexIntersectionsInUnions();
+    //testComplexIntersectionsInUnions();
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/intersection/test_intersection_type.bal:20]: Please create an issue for this.

**Review by MaryamZi — APPROVED** [review]

**Pull request merged**